ratex/csrc/aten_raf_type.cpp

```diff
@@ -1784,6 +1783,82 @@ at::Tensor LazyNativeFunctions::mm(const at::Tensor& self, const at::Tensor& mat
                      /*weight=*/bridge::raf_backend::GetLtcTensor(mat2)));
 }
 
+at::Tensor LazyNativeFunctions::matmul(const at::Tensor& self, const at::Tensor& other) {
+  LTC_FN_COUNTER("raf::");
+  LazyTensor self_tensor = bridge::raf_backend::GetLtcTensor(self);
+  LazyTensor other_tensor = bridge::raf_backend::GetLtcTensor(other);
+  std::vector<int64_t> a_shape =
+      lazy_tensors::util::ToVector<int64_t>(self_tensor.shape().get().dimensions());
+  std::vector<int64_t> b_shape =
+      lazy_tensors::util::ToVector<int64_t>(other_tensor.shape().get().dimensions());
+  int64_t a_size = a_shape.size();
+  int64_t b_size = b_shape.size();
+  int64_t unit_dim = -1;
+  LazyTensor out;
+
+  // TODO support higher demension conversion to lower dim matmul
+  if (a_size > 4 || b_size > 4) {
+    return FALLBACK_ATEN_OP(matmul, self, other);
+  } else if (a_size > 2 && b_size > 2) {
+    // 4 dim matrix w/ unit dim (m5 support)
+    if (a_size > 3) {
+      // Convert to 3D by squeezing unit dim, fallback if cannot squeeze
+      unit_dim = std::find(a_shape.begin(), a_shape.end(), 1) - a_shape.begin();
+      if (unit_dim == a_size)
+        return FALLBACK_ATEN_OP(matmul, self, other);  // No unit dim - unsupported shape
+      self_tensor = LazyTensor::squeeze(self_tensor, unit_dim);
+    }
+    if (b_size > 3) {
+      // Convert to 3D by squeezing unit dim, fallback if cannot squeeze. Ensure if both a & b 4 dim
+      // have unit dim on same axis
+      auto b_dim = std::find(b_shape.begin(), b_shape.end(), 1) - b_shape.begin();
+      if (b_dim != b_size) {
+        if (unit_dim == -1) unit_dim = b_dim;
+        if (b_dim != unit_dim)
+          return FALLBACK_ATEN_OP(matmul, self, other);  // Cannot broadcast - unsupported currently
```

Comment:
It is a little bit confused about what's the current fallback condition. Can you move the conditions into a single if ?
For example,
if ( case1 || case2 || case3 ... ) {
   FALLBACK_ATEN_OP(matmul, self, other);
}